### PR TITLE
GCS Source and Sink additional test cases

### PR DESCRIPTION
**Pull request by sumitsri12**
@sawantpritam Please review

### CODE REVIEW

**Comment by sawantpritam** [comment]
1. Please add proper commit message

2. Your branch is 4 commits behind sawantpritam:e2e-gcs. Please update

File: src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSSource.java [hunk 1]
```diff
@@ -85,4 +85,19 @@ public void enterTheGCSSourceMandatoryProperties() throws InterruptedException,
     CdfGcsActions.getSchema();
     SeleniumHelper.waitElementIsVisible(CdfGCSLocators.getSchemaLoadComplete, 20L);
   }
+
+  @Then("Select file encoding Type {string}")
+  public void selectFileEncodingType(String encoding) {
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSSource.java]: -> Select GCS source property file encoding type

File: src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSSource.java [hunk 2]
```diff
@@ -85,4 +85,19 @@ public void enterTheGCSSourceMandatoryProperties() throws InterruptedException,
...
+    CdfGcsActions.selectFileEncoding(encoding);
+  }
+
+  @Then("Select GCS property read file recursive with {string}")
+  public void selectGCSPropertyReadFileRecursive(String value) {
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSSource.java]: -> Select GCS source property read file recursive as {string}

File: src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSSource.java [hunk 3]
```diff
@@ -85,4 +85,19 @@ public void enterTheGCSSourceMandatoryProperties() throws InterruptedException,
...
+    CdfGcsActions.clickRecursive(value);
+  }
+
+  @Then("Select GCS source property path filename only {string}")
+  public void selectGCSSourcePropertyPathFilenameOnly(String value) {
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSSource.java]: -> Select GCS source property path filename only as  {string}

File: src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSSink.java [hunk 4]
```diff
@@ -81,4 +81,24 @@ public void validateTheCmekKeyOfTargetGCSBucketIfCmekIsEnabled(String cmek) thro
...
+  public void enterGCSSinkPropertyOutputFilePrefix(String prefix) {
+    CdfGcsActions.enterOutputFilePrefix(PluginPropertyUtils.pluginProp(prefix));
+  }
+
+  @Then("Enter GCS sink property path suffix {string}")
+  public void enterGCSSinkPropertyPathSuffix(String pathSuffix) {
+    CdfGcsActions.enterPathSuffix(PluginPropertyUtils.pluginProp(pathSuffix));
+  }
+
+  @Then("Select GCS sink property toggle write header true")
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSSink.java]: ->Toggle GCS sink property write header to true

File: src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSBase.java [hunk 5]
```diff
@@ -184,4 +184,73 @@ public void verifyDataIsTransferredToTargetGCSBucket() {
...
+    String gcsBucket = TestSetupHooks.gcsTargetBucketName;
+    if (verifyFilePresentInGcsBucket(gcsBucket, "." + fileFormat)) {
+      BeforeActions.scenario.write("Data transferred to gcs bucket " + gcsBucket + " with format "
+                                     + fileFormat + " successfully");
+    } else {
+      Assert.fail("Data not transferred to target gcs bucket " + gcsBucket + " with format " + fileFormat);
+    }
+  }
+
+  @Then("Verify data is transferred to target GCS bucket with fileName {string}")
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSBase.java]: -> Verify data is transferred to target GCS bucket with fileName prefix

File: src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSBase.java [hunk 6]
```diff
@@ -184,4 +184,73 @@ public void verifyDataIsTransferredToTargetGCSBucket() {
...
+  public void verifyDataIsTransferredToTargetGCSBucketWithFileNamePrefix(String fileName) {
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSBase.java]: (String fileNamePrefix)

File: src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSBase.java [hunk 7]
```diff
@@ -184,4 +184,73 @@ public void verifyDataIsTransferredToTargetGCSBucket() {
...
+    String gcsBucket = TestSetupHooks.gcsTargetBucketName;
+    String file = PluginPropertyUtils.pluginProp(fileName);
+    if (verifyFilePresentInGcsBucket(gcsBucket, file)) {
+      BeforeActions.scenario.write("Data transferred to gcs bucket " + gcsBucket + " with fileNameContains "
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSBase.java]: fileNameContains -> file name prefix

File: src/e2e-test/features/gcs/sink/GCSSink.feature [hunk 8]
```diff
@@ -29,3 +29,118 @@ Feature: GCS sink - Verification of GCS Sink plugin
...
+    Then Select GCS property format "csv"
+    Then Enter GCS sink property output file prefix "gcsOutputFilePrefix"
+    Then Validate "GCS" plugin properties
+    Then Close the GCS properties
+    Then Save and Deploy Pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Verify data is transferred to target GCS bucket with fileName "gcsOutputFilePrefix"
```
Comment (sawantpritam) [inline src/e2e-test/features/gcs/sink/GCSSink.feature]: -> Verify data is transferred to target GCS bucket with fileName prefix "gcsOutputFilePrefix"

File: src/e2e-test/features/gcs/sink/GCSSink.feature [hunk 9]
```diff
@@ -29,3 +29,118 @@ Feature: GCS sink - Verification of GCS Sink plugin
...
+    Then Select GCS property format "csv"
+    Then Enter GCS sink property path suffix "gcsPathSuffix"
+    Then Validate "GCS" plugin properties
+    Then Close the GCS properties
+    Then Save and Deploy Pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Verify data is transferred to target GCS bucket with fileName "gcsPathSuffix"
```
Comment (sawantpritam) [inline src/e2e-test/features/gcs/sink/GCSSink.feature]: -> Verify data is transferred to target GCS bucket with fileName suffix "gcsPathSuffix"

File: src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSBase.java [hunk 10]
```diff
@@ -184,4 +184,73 @@ public void verifyDataIsTransferredToTargetGCSBucket() {
...
+                                     + fileName + " successfully");
+    } else {
+      Assert.fail("Data not transferred to target gcs bucket " + gcsBucket +
+                    " with fileNameContains " + fileName);
+    }
+  }
+
+  private boolean verifyFilePresentInGcsBucket(String gcsBucket, String fileNameContains) {
+    try {
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSBase.java]: Add one more parameter to this method  String filePart  and 
change if condition based on filePart
if (blob.getName().contains(fileNameContains)) {
          return true;
        }

if filePart is "suffix" -> .endsWith
else if filePart is "prefix" -> .startsWith
else if filePart is "contains" -> .contains
else .equals

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 11]
```diff
@@ -214,4 +222,45 @@ public static void deleteTargetPubSubTopic() {
       }
     }
   }
+
+  @Before(order = 1, value = "@GCS_DATATYPE_1_TEST")
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Add javadoc with list of datatypes

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 12]
```diff
@@ -214,4 +222,45 @@ public static void deleteTargetPubSubTopic() {
...
+  public static void createBucketWithDataTypeTest1File() throws IOException, URISyntaxException {
+    gcsSourceBucketName = createGCSBucketWithFile(PluginPropertyUtils.pluginProp("gcsDataTypeTest1File"));
+  }
+
+  @Before(order = 1, value = "@GCS_DATATYPE_2_TEST")
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Add javadoc with list of datatypes

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 13]
```diff
@@ -214,4 +222,45 @@ public static void deleteTargetPubSubTopic() {
...
+  public static void createBucketWithDataTypeTest2File() throws IOException, URISyntaxException {
+    gcsSourceBucketName = createGCSBucketWithFile(PluginPropertyUtils.pluginProp("gcsDataTypeTest2File"));
+  }
+
+  @Before(order = 1, value = "@GCS_READ_RECURSIVE_TEST")
+  public static void createBucketWithRecursiveTestFiles() throws IOException, URISyntaxException {
+    gcsSourceBucketName = createGCSBucketWithMultipleFiles(PluginPropertyUtils.pluginProp("gcsReadRecursivePath"));
+  }
+
+  @Before(order = 1, value = "@GCS_Parquet_TEST")
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: @GCS_Parquet_TEST -> @GCS_PARQUET_TEST

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 14]
```diff
@@ -214,4 +222,45 @@ public static void deleteTargetPubSubTopic() {
...
+  public static void createBucketWithParquetTestFile() throws IOException, URISyntaxException {
+    gcsSourceBucketName = createGCSBucketWithFile(PluginPropertyUtils.pluginProp("gcsParquetFile"));
+  }
+
+  @Before(order = 1, value = "@GCS_Avro_TEST")
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Tag all CAPS

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 15]
```diff
@@ -75,7 +81,9 @@ public static void createBucketWithOutputFieldFile() throws IOException, URISynt
   }
 
   @After(order = 1, value = "@GCS_CSV_TEST or @GCS_TSV_TEST or @GCS_BLOB_TEST " +
-    "or @GCS_DELIMITED_TEST or @GCS_TEXT_TEST or @GCS_OUTPUT_FIELD_TEST")
+    "or @GCS_DELIMITED_TEST or @GCS_TEXT_TEST or @GCS_OUTPUT_FIELD_TEST or @GCS_DATATYPE_1_TEST or " +
+    "@GCS_DATATYPE_2_TEST " +
+    "or @GCS_READ_RECURSIVE_TEST or @GCS_Parquet_TEST or @GCS_Avro_TEST")
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: TAG ALL CAPS

File: src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature [hunk 16]
```diff
@@ -0,0 +1,212 @@
+@GCS_Source
+Feature: GCS source - Verification of GCS to BQ successful data transfer
+
+  @GCS_CSV_TEST @BQ_SINK_TEST @GCS_READ_RECURSIVE_TEST
```
Comment (sawantpritam) [inline src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature]: remove @GCS_CSV_TEST

File: src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+    Then Verify preview output schema matches the outputSchema captured in properties
+    Then Close the preview data
+    Then Deploy the pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Get count of no of records transferred to target BigQuery Table
+
+  @GCS_CSV_TEST @BQ_SINK_TEST @GCS_DATATYPE_1_TEST
```
Comment (sawantpritam) [inline src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature]: remove @GCS_CSV_TEST 
same comment for all scenarios... remove tags if not required for that scenario

File: src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+    Then Validate "BigQuery" plugin properties
+    Then Close the BigQuery properties
+    Then Save and Deploy Pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Get count of no of records transferred to target BigQuery Table
+
+  @GCS_CSV_TEST @BQ_SINK_TEST @encoding
```
Comment (sawantpritam) [inline src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature]: what is @encoding

File: src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+    Then Validate "BigQuery" plugin properties
+    Then Close the BigQuery properties
+    Then Save and Deploy Pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Get count of no of records transferred to target BigQuery Table
+    Examples:
+      | encodingType |
```
Comment (sawantpritam) [inline src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature]: encodingType -> EncodingType

**Review by sawantpritam — CHANGES_REQUESTED** [review]
@sumitsri12

File: src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSBase.java (hunk 5 above)
Comment (sumitsri12) [reply src/e2e-test/java/io/cdap/plugin/gcs/stepsdesign/GCSBase.java]: I am using this method in both prefix and suffix

**Comment by sumitsri12** [comment]
@sawantpritam  pushed after changes

File: src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+    Then Validate "BigQuery" plugin properties
+    Then Close the BigQuery properties
+    Then Save and Deploy Pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Get count of no of records transferred to target BigQuery Table
+
+
```
Comment (sawantpritam) [inline src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature]: remove extra line

File: src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature [hunk 21]
```diff
@@ -0,0 +1,211 @@
+@GCS_Source
```
Comment (sawantpritam) [inline src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature]: Rename file name - > GCSSourceToBigQueryAdditionalTests.feature

File: src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+    Then Close the BigQuery properties
+    Then Save and Deploy Pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Get count of no of records transferred to target BigQuery Table
+    Then Verify output field "gcsPathField" in target BigQuery table contains filename of the source GcsBucket "gcsOutputFieldTestFile"
+
+
```
Comment (sawantpritam) [inline src/e2e-test/features/gcs/source/GCSSourceAdditionalCases.feature]: remove extra line

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 23]
```diff
@@ -214,4 +222,51 @@ public static void deleteTargetPubSubTopic() {
       }
     }
   }
+
+ /** create gcs bucket with csv file containing below datatypes.
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: -> Create GCS 

Also check formatting.... * on each line

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 24]
```diff
@@ -214,4 +222,51 @@ public static void deleteTargetPubSubTopic() {
...
+  int, float, string, double, long, timestamp, date, boolean, null, bytes
+  */
+  @Before(order = 1, value = "@GCS_DATATYPE_1_TEST")
+  public static void createBucketWithDataTypeTest1File() throws IOException, URISyntaxException {
+    gcsSourceBucketName = createGCSBucketWithFile(PluginPropertyUtils.pluginProp("gcsDataTypeTest1File"));
+  }
+
+  /** create gcs bucket with csv file containing below datatypes.
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Similar comment as above

File: src/e2e-test/resources/pluginParameters.properties [hunk 25]
```diff
@@ -33,6 +78,23 @@ gcsPathField=outputfield
...
+  {"key":"createddate","value":"string"},{"key":"revenue","value":"string"},{"key":"points","value":"string"},\
+  {"key":"longdatatype","value":"string"},{"key":"doubledatatype","value":"double"},\
+  {"key":"date","value":"string"},{"key":"null","value":"string"},{"key":"BytesData","value":"string"}]
+gcsDataTypeTest2File=testdata/GCS_DATATYPE_TEST_2.csv
+gcsDataTypeTest2FileSchema=[{"key":"id","value":"int"},{"key":"name","value":"string"},\
+  {"key":"float_inf","value":"string"},{"key":"float_nan","value":"string"},{"key":"decimal","value":"double"}]
+gcsReadRecursivePath=testdata/GCS_RECURSIVE_TEST
+gcsReadRecursivePathSchema=[{"key":"id","value":"int"},{"key":"name","value":"string"},\
+  {"key":"yearofbirth","value":"int"},{"key":"tablename","value":"string"},{"key":"offset","value":"string"}]
+gcsParquetFile=testdata/GCS_Parquet_TEST.parquet
```
Comment (sawantpritam) [inline src/e2e-test/resources/pluginParameters.properties]: This should match with filename in testdata folder

**Review by sawantpritam — CHANGES_REQUESTED** [review]

File: src/e2e-test/java/io/cdap/plugin/gcs/runners/sourcerunner/TestRunner.java [hunk 26]
```diff
@@ -28,11 +28,14 @@
...
+  /* TODO :Enable tests once issues fixed https://cdap.atlassian.net/browse/PLUGIN-823,
+      https://cdap.atlassian.net/browse/PLUGIN-1113,
+  https://cdap.atlassian.net/browse/PLUGIN-825, https://cdap.atlassian.net/browse/CDAP-18494  */
   monochrome = true,
   plugin = {"pretty", "html:target/cucumber-html-report/gcs-source",
     "json:target/cucumber-reports/cucumber-gcs-source.json",
     "junit:target/cucumber-reports/cucumber-gcs-source.xml"}
 )
-public class TestRunner {
+public class TestRunner1 {
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/gcs/runners/sourcerunner/TestRunner.java]: Rename

File: src/e2e-test/java/io/cdap/plugin/gcs/runners/sinkrunner/TestRunner.java [hunk 27]
```diff
@@ -27,11 +27,12 @@
...
-  tags = {"@GCS_Sink"},
+  tags = {"@GCS_Sink and not @PLUGIN-808"},
+  /* TODO :Enable tests once issue fixed https://cdap.atlassian.net/browse/PLUGIN-808 */
   monochrome = true,
   plugin = {"pretty", "html:target/cucumber-html-report/gcs-sink",
     "json:target/cucumber-reports/cucumber-gcs-sink.json",
     "junit:target/cucumber-reports/cucumber-gcs-sink.xml"}
 )
-public class TestRunner {
+public class TestRunner1 {
```
Comment (sawantpritam) [inline src/e2e-test/java/io/cdap/plugin/gcs/runners/sinkrunner/TestRunner.java]: Rename

**Review by sawantpritam — CHANGES_REQUESTED** [review]

**Review by sawantpritam — APPROVED** [review]

**Comment by sawantpritam** [comment]
Squash the commits.

**Pull request closed without merging**